### ISSUE TITLE
NoReturn doesn't get propagated from function stored in instance variable to method calling said function

### ISSUE THREAD

**Issue by alejo90** [microsoft/pyright#1562]
**Describe the bug**
I have a class that receives a function "returning" `NoReturn` in the constructor and assigns it to a field. Then a method that "returns" no `NoReturn` calls the function assigned to the field. Pyright does identify the type of the field correctly but the method calling the function stored in the field shows an error saying `Function with declared type of "NoReturn" cannot return None`.

**To Reproduce**
Try the snippet below in vscode + Pylance 2021.2.4.

**Expected behavior**
Pyright would understand that the call to the function it identifies as "returning" `NoReturn` implies the method calling it is also "returning" `NoReturn` and not `None`.

**Screenshots or Code**

```python
from typing import Callable, NoReturn


class Foo:
    def __init__(self, f: Callable[..., NoReturn]):
        self.f = f

    def g(self) -> NoReturn:
        self.f()
```

**VS Code extension or command-line**
I'm using vscode + Pylance 2021.2.4 and the issue can also be reproduced in pyright 1.1.117 in basic type checking mode.

![image](https://user-images.githubusercontent.com/1158694/109745164-135ce580-7b99-11eb-9de9-3d22db5496bf.png)

In the image below you can see that pyright does correctly identify the function as "returning" `NoReturn`.

![image](https://user-images.githubusercontent.com/1158694/109745229-2bcd0000-7b99-11eb-8aa6-bd471f578168.png)

**Additional context**
The error is not reported when the type hint is explicitly added to the field assignment in the constructor:

```python
from typing import Callable, NoReturn


class Foo:
    def __init__(self, f: Callable[..., NoReturn]):
        self.f: Callable[..., NoReturn] = f

    def g(self) -> NoReturn:
        self.f()
```

![image](https://user-images.githubusercontent.com/1158694/109745788-01c80d80-7b9a-11eb-9130-5ba698ec1e79.png)

**Comment by erictraut** [comment]
This is a tricky one to fix because the NoReturn detection affects the code flow graph, which in turn affects type inference. The NoReturn detection logic therefore is very limited in its ability to do type inference.

I don't see a straightforward way to fix this particular case. This is an odd edge case, and there's a simple workaround — add a type annotation for the instance variable so its type doesn't need to be inferred.

```python
        self.f: Callable[..., NoReturn] = f
```

**Comment by erictraut** [comment]
I'm going to mark this as "as designed". Let me know if the suggested code change doesn't work for some reason.

**Comment by erictraut** [comment]
I've added more logic to handle this particular case, so this will be addressed in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.199, which I just published. It will also be included in the next release of pylance.
